### PR TITLE
Flos fix

### LINKED ISSUE #7146: Fine tune with local model raised `torch.nn.modules.module.ModuleAttributeError: 'DataParallel' object has no attribute 'config'`
# ❓ Questions & Help

I've fined tune distilgpt2 with `run_language_modeling.py` under my local `output_dir` and want to fine tune with the model in `output_dir` again, but it raised `torch.nn.modules.module.ModuleAttributeError: 'DataParallel' object has no attribute 'config'`. 

## Details

I've fined tune distilgpt2 with `run_language_modeling.py` as follows:
```shell
python3 run_language_modeling.py     --output_dir=/home/xxx/gpt_model/transformers/examples/language-modeling/output_dir --model_type=gpt2  --model_name_or_path=distilgpt2 --per_device_train_batch_size=10     --do_train     --train_data_file=/home/xxx/gpt_model/data_info/data.txt --block_size=64 --save_steps=100 --overwrite_output_dir
```
It works fine and i can get the fine-tuned model after training with my data `data.txt`. Now i want to fine-tune in the `output_dir` so i run `run_language_modeling.py` as follows on my `new_data.txt`:
```shell
 python3 run_language_modeling.py     --output_dir=/home/xxx/gpt_model/transformers/examples/language-modeling/output_new_data --model_type=gpt2  --model_name_or_path=/home/xxx/gpt_model/transformers/examples/language-modeling/output_dir/  --per_device_train_batch_size=20  --do_train     --train_data_file=/home/xxx/gpt_model/data_info/new_data.txt --block_size=64 --save_steps=1000 --overwrite_output_dir
```
But it raise exception and exit. The stderr is as follows
```shell
/home/xxx/gpt_model/pytorch/pytorch/torch/optim/lr_scheduler.py:235: UserWarning: Please also save or load the state of the optimizer when saving or loading the scheduler.
  warnings.warn(SAVE_STATE_WARNING, UserWarning)
Traceback (most recent call last):
  File "run_language_modeling.py", line 320, in <module>
    main()
  File "run_language_modeling.py", line 284, in main
    trainer.train(model_path=model_path)
  File "/home/xxx/anaconda3/envs/transformers/lib/python3.6/site-packages/transformers/trainer.py", line 683, in train
    self.total_flos = getattr(model.config, "total_flos", 0)
  File "/home/xxx/gpt_model/pytorch/pytorch/torch/nn/modules/module.py", line 779, in __getattr__
    type(self).__name__, name))
torch.nn.modules.module.ModuleAttributeError: 'DataParallel' object has no attribute 'config'
```
##### (1) I'm training on multiple GPUs, but i did'n specify a specific gpu to run. So at first i think this may caused by multiple GPUs, and add the following code under https://github.com/huggingface/transformers/blob/52d250f6aa14844024806e5e4dd1c7882bbd8dd5/src/transformers/trainer.py#L641
```shell
if isinstance(model,torch.nn.DataParallel):
		model = model.module
```
It has no error but doesn't work. The output is as follows . After that, the process is killed and return code `echo $?` is `0`.
```shell
Epoch: 0it [00:00, ?it/s]
/home/lenajin/anaconda3/envs/transformers/lib/python3.6/site-packages/transformers/trainer.py:1087: FutureWarning: This method is deprecated, use `Trainer.is_world_process_zero()` instead.
  warnings.warn("This method is deprecated, use `Trainer.is_world_process_zero()` instead.", FutureWarning)
```
##### (2) Following with https://github.com/huggingface/transformers/issues/1991 , I try to train with following script to train with gpu whose id is `1`.
```shell
CUDA_VISIBLE_DEVICES=1 python3 run_language_modeling.py     --output_dir=/home/xxx/gpt_model/transformers/examples/language-modeling/output_dir --model_type=gpt2  --model_name_or_path=distilgpt2 --per_device_train_batch_size=20     --do_train     --train_data_file=/home/xxx/gpt_model/data_info/data.txt --block_size=64 --save_steps=100 --overwrite_output_dir
```
it return with return code `echo $?` =`0`. At first i don't know what's the problem, but now i changed `--per_device_train_batch_size=20    ` from 20 to 10. It seems all is well, and now my `GPU-Util` is about `98%`.  Maybe it return without training duing to my limit gpu CC？It's wired but at least it works now. Maybe there can give some error message about the return reason?

### PR DESCRIPTION
**Pull request by marrrcin**
Fixes #7146 

This basically unwraps the model that is used during training and can be either plain `Module` or `DataParallel`/`DistributedDataParallel`.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1448,15 +1448,29 @@ def floating_point_ops(self, inputs: Dict[str, Union[torch.Tensor, Any]]):
...
 
         if hasattr(model, "floating_point_ops"):
             return model.floating_point_ops(inputs)
 
         else:
             return 0
+
+    @staticmethod
+    def _actual_model(
+        model: Union[torch.nn.DataParallel, torch.nn.parallel.DistributedDataParallel, torch.nn.modules.Module]
```
Comment (sgugger) [inline src/transformers/trainer.py:1461]: nit: aren't them all subclasses of `torch.nn.Module`?

**Review by sgugger — APPROVED** [review]
Thanks a lot for the quick fix! This LGTM.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1448,15 +1448,29 @@ def floating_point_ops(self, inputs: Dict[str, Union[torch.Tensor, Any]]):
...
+        model = self._actual_model(self.model)
 
         if hasattr(model, "floating_point_ops"):
             return model.floating_point_ops(inputs)
 
         else:
             return 0
+
+    @staticmethod
+    def _actual_model(
```
Comment (TevenLeScao) [inline src/transformers/trainer.py:1460]: The same logic is used in `floating_point_ops` - could you also change it to use `_actual_model`?

File: tests/test_trainer.py [hunk 3]
```diff
@@ -336,3 +336,16 @@ def test_num_train_epochs_in_training(self):
         trainer = get_regression_trainer(train_len=64, per_device_train_batch_size=16, gradient_accumulation_steps=5)
         train_output = trainer.train()
         self.assertEqual(train_output.global_step, int(self.n_epochs))
+
+    def test_flos_extraction(self):
```
Comment (TevenLeScao) [inline tests/test_trainer.py:340]: Thanks for adding the test!

File: src/transformers/trainer.py (hunk 2 above)
Comment (marrrcin) [reply src/transformers/trainer.py:1460]: Hmm... It is already changed in line 1451 🤔

File: src/transformers/trainer.py (hunk 1 above)
Comment (marrrcin) [reply src/transformers/trainer.py:1461]: I wanted to be explicit, but I can change it if you think that's required.

**Comment by marrrcin** [comment]
Please merge at will, as this fix is blocking us (https://github.com/huggingface/transformers/issues/7146#issuecomment-698852274).

**Review by LysandreJik — APPROVED** [review]

**Pull request merged**